### DISCUSSION TITLE
mac os packetflix hander and xattr

### DISCUSSION THREAD

**Discussion by hellt** [siemens/edgeshark#46] [General]
I am not sure what caused the packetflix-handler.zip to be detected as a malicious download
First, chrome warns a user that this is a malicious download, and then suddenly xattr is required.

None of these steps were necessary for the handler app I initially shared, so I think it makes sense to dig a little bit deeper and see what caused that behavior.  https://github.com/siemens/cshargextcap/issues/14

And maybe there is a wireshark option that sets the window title that can also be set in the AppleScript to actually show the contianer/pid name and interface for the captured flow. Right now it is just saying "capturing from packetflix"

<img width="693" alt="image" src="https://github.com/siemens/edgeshark/assets/5679861/fd6ebd3e-35d5-4269-9cfc-69bdd69aaaab">

**Comment by thediveo** [comment]
What I observed and what killed a few hours of my lifetime: when I zipped everything on the same machine and then unzipped again in Downloads/ and copied to /Applications there was never an issue.

But downloading triggers the whole (cyber) security cargo culture we all so love. Luckily, I remembered that we just lost a few hours worth of company time on another download quarantine a few weeks ago and luckily someone then remembered that there is an xattr flag. Took some more time to finally nail down the correct flag.

**Reply by hellt** [reply]
I doubt that is the case...
Have you tried my zip package? People used it on different macs and they never had those ugly warnings (besides the regular unsigned app warning), hence they never used xattr command either

**Reply by thediveo** [reply]
In our security guidelines I cannot try that. In the end, there must be a clear source of everything that ends up in the repo. How did your testers access the zip package?

**Reply by hellt** [reply]
they downloaded it from the comment where I first introduced it.
My point is, there is something "strange" with contents of your apple script that causes these unexpected warnings, and I'd love to know what and why.
Especially given that my archive doesn't exhibit the same warnings

**Reply by thediveo** [reply]
https://forums.developer.apple.com/forums/thread/666452?answerId=647110022#647110022 from apple's developer forum...

![use curl](https://github.com/siemens/edgeshark/assets/6920158/c6d33a93-79cd-4d2e-b39b-e2bc5c1c3ec7)

maybe we should post a "I feel lucky" one-line in the same spirit as the curl the compose case?

Finally, an macos OpenSource installer ... ash script 🤦🏼

**Reply by thediveo** [reply]
let's see:
1. macos Monterey 12.7.2
2. Script Editor 2.11 (227)
3. zip says 3.0 July 5th 2008
4. Firefox -- not Chrome -- 122.0.1 for downloading

**Reply by thediveo** [reply]
same with Chrome 121.0.6167.160.

**Reply by hellt** [reply]
Let me try downloading via curl indeed.
If it doesn't quarantine, then I believe this should be the official way to download the file.

I propose using the following oneliner:

```bash
curl -s https://api.github.com/repos/siemens/cshargextcap/releases/latest \
  | grep "browser_download_url.*packetflix-handler.zip" \
  | cut -d : -f 2,3 | tr -d \" \
  | xargs -n 1 curl -sLO
```

**Reply by hellt** [reply]
I just tried using curl and no security warnings occurred 🥳 
maybe then it is worth adding `&& unzip packetflix-handler.zip && mv packetflix-handler.zip /Applications` to the oneliner to make it a really one-click install

**Reply by thediveo** [reply]
what if an older downloaded version is still in the Downloads? does curl overwrite or else...?

**Comment by thediveo** [comment]
re: title; looking at https://www.wireshark.org/docs/wsug_html_chunked/ChCustCommandLine.html, there isn't anything available from the wireshark CLI args. The extcap definitively hasn't access to this. I would recommend hopping over to Wireshark's discord server and into the "developer den", asking there for this feature.

**Reply by hellt** [reply]
We can use the custom option capability:

```
/Applications/Wireshark.app/Contents/MacOS/Wireshark -o gui.window_title:hello-edgeshark
```

there is also `gui.start_title` that is by default set to "The World's Most Popular Network Protocol Analyzer"

**Reply by thediveo** [reply]
that leaves us still with the problem that we're giving only an URL and how to get the title off the query part of the URL, and this on three different platforms...

**Reply by hellt** [reply]
I think I can do that for MacOS,
In the applescript we have everything we need - remote host addr, container-name and if-name.

If chatgpt knows applescript syntax then it can parse those elements from it. I will check this

**Reply by thediveo** [reply]
problem for me is to have this also for the other platforms as well

**Reply by hellt** [reply]
**ADD1:** ooops, wrong copypaste, working on a proper paste
**ADD2:** code contents fixed

would you be willing to change the MacOS script to feature the following:

```
on open location thisURL
	set unescURL to do shell script "echo " & quoted form of thisURL & " | python3 -c 'import sys, urllib.parse as ul; print(ul.unquote_plus(sys.stdin.read()))'"
	set netIFs to do shell script "echo " & quoted form of unescURL & " | awk -F'container=' '{print $2}' | python3 -c \"import sys, json; data = json.load(sys.stdin); print('{}:{}'.format(data['name'],data['network-interfaces']))\""
	do shell script "/Applications/Wireshark.app/Contents/MacOS/Wireshark -k -i packetflix -o gui.window_title:\"" & netIFs & "\" -o extcap.packetflix.url:\"" & thisURL & "\""
end open location
```

It uses python3 that should be available on non-ancient macs to parse the json in the URL to extract the name and interface list.

Here is how it works:

https://github.com/siemens/edgeshark/assets/5679861/b9b051b4-bd82-4db6-b967-c0c4ef4d8dee

**Reply by thediveo** [reply]
are you aware of this? https://stackoverflow.com/a/73360794/6632214 for parsing and decoding URLs in Apple Script? Maybe we can avoid the py3 dependeny?

**Reply by hellt** [reply]
I am not sure that would help this case... We can get the path and query with that, but the query string that embeds the json still needs to be parsed

**Reply by thediveo** [reply]
ChatGPT, when massaged, knows something about NSJSONSerialization and dicts that seem to do the trick. At first, it resorted to awk and sed, which doesn't seem so bad either.

**Reply by hellt** [reply]
ah, that would be sweet indeed. I didn't _dive_ that deep...

**Reply by thediveo** [reply]
wondering whether we get also something in powershell 5, and some bash magic in linux should be more straightforward.

**Reply by hellt** [reply]
you can add me as a co-author in the commit :D

```
use framework "Foundation"
use scripting additions

on open location thisURL
	set theURL to current application's NSURL's URLWithString:thisURL
	
	set theComponents to current application's NSURLComponents's componentsWithURL:theURL resolvingAgainstBaseURL:false
	set queryItems to theComponents's queryItems
	set containerValue to ""
	
	set theEnumerator to queryItems's objectEnumerator()
	set aQueryItem to theEnumerator's nextObject()
	
	repeat while aQueryItem is not missing value
		if ((aQueryItem's valueForKey:"name") as string is "container") then
			set containerValue to ((aQueryItem's valueForKey:"value") as string)
			exit repeat
		end if
		set aQueryItem to theEnumerator's nextObject()
	end repeat
	-- uncomment to see the parsed json from the URL
	--	display dialog (containerValue as string)
	
	
	try
		-- Convert the containerValue to NSData
		set nsContainerValue to current application's NSString's stringWithString:containerValue
		set theData to nsContainerValue's dataUsingEncoding:(current application's NSUTF8StringEncoding)
		
		-- Parse the containerValue as JSON
		set theDict to current application's NSJSONSerialization's JSONObjectWithData:theData options:0 |error|:(missing value)
		
		-- Get the value of the namespace/container "name" key
		set nodeName to (theDict's valueForKey:"name") as string
		
		-- Get the value of the "network-interfaces" key
		set networkInterfacesList to (theDict's valueForKey:"network-interfaces") as list
		
		-- Concatenate the elements into a comma-separated string
		set netIFs to ""
		repeat with i from 1 to count networkInterfacesList
			set netIFs to netIFs & item i of networkInterfacesList
			if i is not (count networkInterfacesList) then
				set netIFs to netIFs & ", "
			end if
		end repeat
		
		-- Display the value in a dialog box
		--		display dialog (netIFs as string)
		--		display dialog (nodeName as string)
		-- open wireshark
		-- create a title for wireshark window
		set title to nodeName & ": " & netIFs
	on error errMsg number errNum
		display dialog ("Error " & errNum & ": " & errMsg)
	end try
	do shell script "/Applications/Wireshark.app/Contents/MacOS/Wireshark -k -i packetflix -o gui.window_title:\"" & title & "\" -o extcap.packetflix.url:\"" & thisURL & "\""
end open location
```

I hate it more than py because there is no way I can remember what happened here in a week from now, but maybe it is better then shelling out to py3

**Reply by thediveo** [reply]
Here we are, with Apple's (well, NeXT's) very own COBOL dialect.

**Reply by thediveo** [reply]
fun fact: both Ghostwire and Packetflix started their life as Python3 implementations. I then soon rewrote Packetflix in Go because of better performance, memory consumption, and especially maintenance (single binary, not the whole Python ecosystem in a container anymore). Ghostwire took much longer, and only after my personal experiment lxkns showed that it is possible to write a Linux kernel namespace discovery engine in Go.

**Reply by thediveo** [reply]
released in v0.10.7. Thank you very much, @hellt for your valuable contribution! Of course, you're mentioned as co-author, that's the least I can do.

Now I might have an idea how to get the same feature on Linux and Windows ... maybe I can reuse the extcap binary as a shim loader for Wireshark...
